02-peer-protocol.md

```diff
@@ -197,6 +206,13 @@ know this node will accept `funding_satoshis` greater than or equal to 2^24.
 Since it's broadcast in the `node_announcement` message other nodes can use it to identify peers 
 willing to accept large channel even before exchanging the `init` message with them. 
 
+Channel features are explicitly enumerated as `channel_type`
+bitfields, using odd features bits.  The currently defined types are:
```

Comment:
Because we already have one mechanism to coordinate, let's not create another!  For example, I already have variants which add option_simplified_update to all of these....

Reply:
> Because we already have one mechanism to coordinate

If we don't want to have to assign a new set of bits, we can just use the current bit position number for the actual feature bits themselves. Since it's just an integer, we don't waste any space by not assigning non-contiguous values. To simplify things a bit the optional bit can be used, so static key is 13, anchor zero htlc is 23, etc, etc. 

The other issue with using the raw feature bit vector here is the size of the message. This likely isn't the last TLV we'll add here, and given it's an embedded feature vector it can swell to several KBs which means the message may run up against the current max message size.

Using a set integer based off the feature bit position addresses all these issues and doesn't attempt to send a redundant set of bytes to communicate things we already know. If we've already sent our feature bits in the node ann, and in innit upon connection, why do we need to send them again here? I already know what you support!  

> what do I do if I receive a byte array contains bits 7, 13, and 21? Or if I receive bits 13, 21 and 37? The first one is harmless, the second one should probably be rejected

This example shows how sending the feature bits again doesn't actually make the negotiation _explicit_ and just adds more feature bit compatibility issues at a second site (first in the node ann, and now here again). Having the acceptor "pick" amongst an advertised set just adds more moving parts, and also makes weird downgrade cases (that we've grappled w/ in the past) exist at another site (what does "most preferred" even mean?). 

If I send type `13` (static key), then it's 100% unambiguous what type of channel I want to open, no extra mapping (set of feature bits to a type) or implicit negotiation needed. It's dead simple. 

> For example, I already have variants which add option_simplified_update to all of these....

This is an argument for just using a single type imo. Lets say CL implement that new commitment update type, when newer more distinct commitment types (witness asymmetric commitments, eltoo, all the taproot variants, hybrid DLC etc) arise, that protocol will likely need to be revamped to support the new HTLC/commitment variants. Rather than needing to interpret that "modification" bit, a _new_ commitment type should be added instead to enumerate all the possible supported update+commitment variants. This makes things _explicit_ and gets rid of the weird (and error prone) implicit negotiation we have atm. This enumeration doesn't add any in-protocol overhead as a `bigsize` integer can be used. 

[We've already implemented it](https://github.com/lightningnetwork/lnd/pull/5373) and just using that type directly is a lot simpler than needing to map the set of feature bits (and determine incompatibilities or ambiguities, etc) to a type that ultimately needs to exist anyway for any implementation.

Reply:
I've thought about this more, and the issue with a `channel_type` enum as done by lnd (and by myself at first) is the explosion of enum values when you combine orthogonal features (**the `channel_type` isn't only about the commitment format, it's really the channel's feature set**).

Right now we have:

- "basic" channels
- static remotekey channels
- anchor output channels
- anchor output with 0-fee htlc channels

If we add two features that are orthogonal to the commitment format, for example `option_simplified_update` and `option_zero_reserve` (these are just examples, imagine two channel features that are completely orthogonal to the commitment format), the enum becomes:

- "basic" channels
- static remotekey channels
- anchor output channels
- anchor output with 0-fee htlc channels
- "basic" channels with simplified update
- static remotekey channels with simplified update
- anchor output channels with simplified update
- anchor output with 0-fee htlc channels with simplified update
- "basic" channels with zero reserve
- static remotekey channels with zero reserve
- anchor output channels with zero reserve
- anchor output with 0-fee htlc channels with zero reserve
- "basic" channels with zero reserve and simplified update
- static remotekey channels with zero reserve and simplified update
- anchor output channels with zero reserve and simplified update
- anchor output with 0-fee htlc channels with zero reserve and simplified update

That simply doesn't scale...and is correctly handled by rusty's proposal IMHO. In the channel DB we'd just store the feature bytes array corresponding to channel features (features that affect the channel parameters/format). As discussed during yesterday's spec meeting, each `channel_type` is uniquely identified by the value of its `features` field so it's not ambiguous.

I don't think the size of the message is a real issue. Even if we have 150 feature bits (which will take decades), it means each `channel_type` is ~20 bytes. If your implementation supports 50 distinct `channel_type`s (which is probably unreasonable from a security point of view) that's only 1kB. It can be an issue if you want to use features that aren't in the spec though, and that need to use a high feature bit. In that case it can be a bit higher, but if you use a feature bit in the 150-300 range it's still not an issue imo.

Each implementation would probably internally define the list of `channel_type`s they want to support, ordered by preference (e.g. I'd like to only support static_remotekey and anchor outputs, with or without option_simplified_update, and I'd prefer using anchor outputs and option_simplified_update) and somehow let the node operator configure what she wants to allow/disallow, but how to present that to the node operator is orthogonal to how we internally negotiate it in the channel open protocol and store it in our channel database.

Reply:
> is the explosion of enum values when you combine orthogonal features

Yeah this is the tradeoff in terms of the "enum space": everything needs to be unrolled. However there isn't any added overhead on the wire given that you can pack just about every combination you can think of into a `bigsize` integer. 

> That simply doesn't scale..

I guess it depends where you think the scale issues arise. It's simply taking up space already allocated (bit wise) in the integer specified. In the end, if you send all the individual feature bits, you'll end up needing to wrangle with the same "parameter space" IMO, as you need to map those down to an actual channel type (concrete implementation).

One alternative would be to split things into a base channel type, and a modifier. In your example, the modifiers would be the zero reserve and the simplified update channel type. 

I think in the end though, all implementations simply won't support every combination of the feature bit expansion (we shouldn't underestimate the complexity here). As an example, lnd doesn't support making a "basic" channel any longer, though we need to _continue_ to signal the optional bit, as otherwise with the way feature bits are set up right now, things break down on the connection level. As more and more channel types are added (taproot variants, eltoo, DLC, etc, etc) IMO it'll become more and more difficult for an implementation to support all of them. Having an "enum" for the channel type simply makes that expansion _explicit_. 

> I don't think the size of the message is a real issue.
> If your implementation supports 50 distinct channel_types (which is probably unreasonable from a security point of view) that's only 1kB

IMO the issue here is that as the TLV expands in this message, it may become impossible to honestly advertise the set of parameters you want, as you risk running into the max message limit. Why send kilobytes when a 4 bytes will do? It's unnecessarily inefficient and made worse by the fact that we don't use a sparse encoding. 

The other issue I see here as mentioned above is that we're just sending the exact same feature bit information that's already present and has been exchanged potentially twice at this point (in the init and node_ann message). On top of this, by letting the _responder_ ultimately select the channel type, the initiator is unable to actually pin down the channel type they want to send, other than sending a single feature bit (or set of bits), which reduces down to simply sending the channel type. 

On top of that, if I want to open an explicit channel type (I only want the "most secure" option), why would I ever send an array of options and allow the receiver to _choose_ amongst them? It's unclear how to advise implementations to create a safe generalized "channel type satisficement" engine (accepting the "best" set of bits). I don't think we should easily dismiss the complexity added here w.r.t preference and selection. 

In short I see future potential for a series of inadvertent footguns introduced by adding a loose "negotiation layer". An implementation needs to wrangle with parsing all the different combinations, and choosing which one is "best" for it (to advertise and accept) vs just sending what they want (based on the feature bits already advertised at the othe layers) and the other side rejecting or accepting.

Reply:
> One alternative would be to split things into a base channel type, and a modifier. In your example, the modifiers would be the zero reserve and the simplified update channel type.

I agree, but these modifiers are exactly the same thing as feature bits, aren't they? So why introduce a different mechanism?

> IMO the issue here is that as the TLV expands in this message, it may become impossible to honestly advertise the set of parameters you want, as you risk running into the max message limit. Why send kilobytes when a 4 bytes will do? It's unnecessarily inefficient and made worse by the fact that we don't use a sparse encoding.

In my example, 1kB is really the most extreme case, I'm pretty sure this will never use more than ~100 bytes in the next couple of years. IMHO the number of bytes used in the messages isn't a strong enough argument to weigh in favor of one proposal or the other.

> I think in the end though, all implementations simply won't support every combination of the feature bit expansion

I completely agree, but I can really see a future where there a few distinct orthogonal feature bits that an implementation may support, and where this mechanism is useful (an implementation could very well support two distinct update protocols, two distinct commitment formats and a few minor "modifiers").

> On top of that, if I want to open an explicit channel type (I only want the "most secure" option), why would I ever send an array of options 

You don't have to. If you want only one `channel_type`, you should send just that one in your message. But if other implementations want to let the fundee choose among a range of options, we should give them the freedom to do that, shouldn't we? Especially once we have some channel features that are orthogonal to the commitment format, the choice of which one to use will be a choice of what trade-offs you'd rather make, so I expect nodes to have different preferences here, hence I think giving them a choice would make sense.